addon/utils/query-cache.js

```diff
@@ -0,0 +1,69 @@
+
+const CACHE_DECAY_TIMEOUT = 5000;
+
+export default class QueryCache {
+  constructor() {
+    this._internalCache = Object.create(null);
+    this.decayId = null;
+  }
+
+  add(type, key, value) {
+    if (!key || !value) {
+      return;
+    }
+
+    if (!this._internalCache[type]) {
+      this._internalCache[type] = Object.create(null);
+    }
+
+    this._internalCache[type][key] = value;
+  }
+
+  remove(type, key) {
+    let typeMap = this._internalCache[type];
+
+    if (typeMap && typeMap[key]) {
+      delete typeMap[key];
+    }
+  }
+
+  clear() {
+    this._internalCache = Object.create(null);
+  }
+
+  get(type, key) {
+    if (!this._internalCache[type]) {
+      return;
+    }
+
+    return this._internalCache[type][key];
+  }
+
+  get types() {
+    return Object.keys(this._internalCache);
+  }
+
+  scheduleDecay(delay = CACHE_DECAY_TIMEOUT, onDecay) {
```

Comment:
Let's back up here for a minute. I think we should start identifying use cases before we move any further. This will allow us to tailor the implementation to what we need *now*, not what we think we'll need in the future. I'm not saying we won't handle future use cases, but we should focus on immediate needs an go from there.